File: metrics/src/tests/countmetrictest.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include <vespa/vespalib/objects/floatingpointtype.h>
#include <vespa/metrics/countmetric.h>
#include <gtest/gtest.h>

using vespalib::Double;

namespace metrics {

TEST(CountMetricTest, testLongCountMetric)
{
    LongCountMetric m("test", {{"tag"}}, "description");
    m.set(100);
    EXPECT_EQ(uint64_t(100), m.getValue());
    m.inc(5);
    EXPECT_EQ(uint64_t(105), m.getValue());
    m.dec(15);
    EXPECT_EQ(uint64_t(90), m.getValue());
    LongCountMetric m2(m);
    EXPECT_EQ(uint64_t(90), m2.getValue());
    m.reset();
    EXPECT_EQ(uint64_t(0), m.getValue());

    LongCountMetric n("m2", {}, "desc");
    n.set(6);
    EXPECT_EQ(uint64_t(6), n.getValue());

    LongCountMetric o = m2 + n;
    EXPECT_EQ(uint64_t(96), o.getValue());

    o = m2 - n;
    EXPECT_EQ(uint64_t(84), o.getValue());

    std::string expected("test count=84");
    EXPECT_EQ(expected, o.toString());

    EXPECT_EQ(Double(84), Double(o.getDoubleValue("value")));
    EXPECT_EQ(int64_t(84), o.getLongValue("value"));
}

}

File: metrics/src/test/java/com/yahoo/metrics/CountMetricTest.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.metrics;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * @author thomasg
 */
public class CountMetricTest {

    @Test
    public void testCountMetric() {
        CountMetric m = new CountMetric("test", "tag", "description", null);
        assertEquals(false, m.used());
        m.set(100);
        assertEquals(true, m.used());
        assertEquals(100, m.getValue());
        m.inc(5);
        assertEquals(105, m.getValue());
        m.dec(15);
        assertEquals(90, m.getValue());

        CountMetric m2 = new CountMetric(m, Metric.CopyType.CLONE, null);
        m.reset();
        assertEquals(90, m2.getValue());
        assertEquals(0, m.getValue());

        CountMetric n = new CountMetric("m2", "", "desc", null);
        n.set(6);

        n.addToSnapshot(m2);
        assertEquals(96, m2.getValue());
        n.addToPart(m);
        assertEquals(6, m.getValue());

        assertEquals(6, n.getValue());

        assertEquals("<test description=\"description\" count=\"96\"/>\n", m2.toXml(0, 2));
        assertEquals("<test description=\"description\" count=\"96\" average_change_per_second=\"9.60\"/>\n", m2.toXml(10, 2));
        assertEquals(96.0, m2.getDoubleValue("value"), 0.00000001);
        assertEquals(96, m2.getLongValue("value"));
    }

}
